File: stock-tracker/src/chart_input/factory.rs
use crate::{
  chart_input::{
    InputDaily, InputIntraday, InputMonthly, InputWeekly, LoadChartInputTrait,
  },
  stock_event::EventPayloadTrackStock,
};

pub fn create_input(
  payload: &EventPayloadTrackStock,
) -> Result<Box<dyn LoadChartInputTrait>, &'static str> {
  let symbol = payload.symbol.clone();
  let interval = payload.interval.clone();
  let chart_type = payload.chart_type.as_str();

  match chart_type {
    "intraday" => Ok(Box::new(InputIntraday::new(symbol, interval))),
    "daily" => Ok(Box::new(InputDaily::new(symbol))),
    "weekly" => Ok(Box::new(InputWeekly::new(symbol))),
    "monthly" => Ok(Box::new(InputMonthly::new(symbol))),
    _ => Err("invalid chart type"),
  }
}

File: chart-api/src/chart-input/factory.ts
import { ChartRequestInput } from "../model";
import { LoadChartInputBase } from './input.base';
import { LoadChartInputIntraday } from "./input.intraday";
import { LoadChartInputDaily } from "./input.daily";
import { LoadChartInputWeekly } from "./input.weekly";
import { LoadChartInputMonthly } from "./input.monthly";

export const toLoadChartInput = (
  reqInput: Readonly<ChartRequestInput>,
): LoadChartInputBase => {
  const { chartType, symbol, interval } = reqInput;

  switch (chartType) {
    case 'intraday': return new LoadChartInputIntraday(symbol, interval as string);
    case 'daily': return new LoadChartInputDaily(symbol);
    case 'weekly': return new LoadChartInputWeekly(symbol);
    case 'monthly': return new LoadChartInputMonthly(symbol);
    default: throw new Error('invalid chart type');
  }
}